README.md
# SwiftCSSBuilder

[![Version](https://img.shields.io/cocoapods/v/SwiftCSSBuilder.svg?style=flat)](https://cocoapods.org/pods/SwiftCSSBuilder)
[![License](https://img.shields.io/cocoapods/l/SwiftCSSBuilder.svg?style=flat)](https://cocoapods.org/pods/SwiftCSSBuilder)
[![Platform](https://img.shields.io/cocoapods/p/SwiftCSSBuilder.svg?style=flat)](https://cocoapods.org/pods/SwiftCSSBuilder)

## Installation

### Using CocoaPods

```ruby
pod 'SwiftCSSBuilder'
```

## Basic Usage

With SwiftCSSBuilder you can build Cascading Style Sheets (CSS) by constructing them in swift language. See the following example:

```swift
let style = CSS.Stylesheet()
    .with(CSS.Style(for: .all)
            .property(.fontFamily, value: "Helvetica")
            .property(.fontSize, value: .pt(20))
            .property(.color, value: "#606060"))
    .with(CSS.Style(for: .class("cuteDiv"))
            .property("background-color", value: .color(UIColor.gray))
            .property(.padding, value: .px(20)))
    .with(CSS.Style(for: .p.childOf(.div.withClass("cuteDiv")))
            .property(.color, value: .color(.red)))
    
print(style.toString())
```

Which outputs the following CSS:

```css
* {
    font-family: Helvetica;
    font-size: 20.0pt;
    color: #606060;
}

.cuteDiv {
    background-color: #000000FF;
    padding: 20px;
}

div.cuteDiv > p {
    color: #FF0000FF;
}

```

You can specify properties by hand by specifing the CSS property name and value:

```swift
.property("margin-bottom", value: "1em")
```

Or you can use one of the build in enums in CSS.PropertyType for the property name, and an enum from CSS.PropertyValue for the value, for example:

```swift
.property(.marginBottom, value: .em(1))
```

Both will translate to the following CSS property:

```css
margin-bottom: 1em;
```

You can even use an hand written property name via String and use a builtin CSS.PropertyValue for the value:

```swift
.property("font-size", value: .pt(12))
```

Or, viceversa:

```swift
.property(.fontSize, value: "12pt")
```

## Passing UIKit types

You can use some UIKit types with SwiftCSSBuilder by passing them to specific CSS.PropertyValue, for example:

You can specify a CSS color property by passing an UIColor instance. SwiftCSSBuilder will take and translate the color from that UIColor instance:

```swift
.property(.color, value: .color(UIColor.red))
```

You can specify a CSS font related property by passing an UIFont instance. SwiftCSSBuilder will take the property value from that UIFont instance:

```swift
let font: UIFont = //A cool UIFont my app uses
.property(.fontFamily, value: .fontFamily(font)) //Takes the font-family from the specified UIFont
.property(.fontSize, value: .fontSize(font)) //Takes the font-size from the specified UIFont
```

You can also specify one or more fallback fonts for the font-family CSS property:

```swift
let font: UIFont = //A cool UIFont my app uses
.property(.fontFamily, value: .fontFamily(font, fallbackFonts: ["Arial", "sans-serif"])) //Takes the font-family from the specified UIFont and specifies Arial and sans-serif as fallback fonts 
```

## Built-in Property Values specifiers

As said in the previous paragraphs, you can specify both hand written properties or you can use the built-in CSS.PropertyType and CSS.PropertyValue enums to specify property names and value. The following is a list of some useful built-in CSS.PropertyValue specifiers:

### .color

Allows to specify a color-related property's value by using a UIColor instance. Example:

```swift
.property(.color, value: .color(UIColor.red))
```

### .RGB / .RGBA / .fRGB, .fRGBA

Allows to specify a color-related property's value by using Red, Green, Blue (and eventually Alpha) components as Int (0-255) or as CGFloat (0.0-1.0). Examples:

```swift
.property(.color, value: .RGB(255, 0, 0))
```

```swift
.property(.color, value: .RGBA(255, 0, 0, 255))
```

```swift
.property(.color, value: .fRGB(1.0, 0.0, 0.0))
```

```swift
.property(.color, value: .fRGBA(1.0, 0.0, 0.0, 1.0))
```

### .fontFamily

Allows to specify a font-family CSS property's value by using an UIFont. Example:

```swift
let font: UIFont = //Your UIFont instance
.property(.fontFamily, value: .fontFamily(font))
```

### .fontSize

Allows to specify a font-size CSS property's value by using an UIFont. Example:

```swift
let font: UIFont = //Your UIFont instance
.property(.fontSize, value: .fontSize(font))
```

### .px, .pt, .em, .perc

Allows to specify a generic CSS properties' values by using px, pt, em and percentual units. Example

```swift
.property(.marginTop, value: .px(10))
```

Translates to:

```css
margin-top: 10px;
```

Or you can specify 2 or 4 values:

```swift
.property(.margin, value: .px(10, 20))
.property(.padding, value: .em(10, 20, 30, 40))
```

Translates to:

```css
margin: 10px 20px;
padding: 10em 20em 30em 40em;
```

```swift
.property("font-size", value: .pt(12))
```

Translates to:

```css
font-size: 12pt;
```

```swift
.property("max-height", value: .perc(50))
```

Translates to:

```css
max-height: 50%;
```

## Advanced CSS selectors

The following properties are applied to all div elements which has the myCuteDiv html class:

```swift
let style = CSS.Stylesheet()
    .with(CSS.Style(for: .div.withClass("myCuteDiv")
            .property(/* Your property */)
            .property(/* Your property */)
```


The following properties are applied to all p elements which are childs of div elements with class "myCuteDiv":

```swift
let style = CSS.Stylesheet()
    .with(CSS.Style(for: .p.childOf(.div.withClass("myCuteDiv")))
            .property(/* Your property */)
            .property(/* Your property */)
```

The following properties are applied to all a elements with :hover pseudoclass:

```swift
let style = CSS.Stylesheet()
    .with(CSS.Style(for: .a.withPseudoClass("hover"))
            .property(/* Your property */)
            .property(/* Your property */)
```

The following properties are applied to all input elements with name attribute equal to "username":

```swift
let style = CSS.Stylesheet()
    .with(CSS.Style(for: .custom("input").withAttribute("name", value: "username"))
            .property(/* Your property */)
            .property(/* Your property */)
```

... And you can obviously combine them together for even complex selectors!

## Example Project

To run the example project, clone the repo, and run `pod install` from the Example directory first.

## Requirements

## Installation

SwiftCSSBuilder is available through [CocoaPods](https://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod 'SwiftCSSBuilder'
```

## Author

Marco Siino, siino.marco@gmail.com

## License

SwiftCSSBuilder is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# CHANGELOG

## 0.1.0

Released the first version of the library



Example/Tests/Tests.swift
import XCTest
import SwiftCSSBuilder

class Tests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        XCTAssert(true, "Pass")
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure() {
            // Put the code you want to measure the time of here.
        }
    }
    
}


SwiftCSSBuilder/Classes/Style.swift
//
//
//  Style.swift
//
//  Created on 03/03/22
//  Copyright (c) 2022 Marco Siino. All rights reserved.
//


import Foundation

extension CSS {
    
    public struct Style {
        private let selectors: [Selector]
        private var properties = [Property]()

        public init(`for` selector: Selector) {
            self.selectors = [selector]
        }
        
        public init(`for` selectors: [Selector]) {
            self.selectors = selectors
        }
        
        public init(`for` selector: Selector, properties: [Property]) {
            self.init(for: selector)
            self.properties = properties
        }
        
        public init(`for` selectors: [Selector], properties: [Property]) {
            self.init(for: selectors)
            self.properties = properties
        }
        
        @discardableResult public func property(_ prop: PropertyType, value: PropertyValue) -> Style {
            var properties = self.properties
            properties.append(Property(property: prop, value: value))
            return Style(for: selectors, properties: properties)
        }
        
        @discardableResult public func property(_ prop: PropertyType, value: String) -> Style {
            return property(prop, value: .custom(value: value))
        }
        
        @discardableResult public func property(_ customProperty: String, value: String) -> Style {
            return property(.custom(customProperty), value: .custom(value: value))
        }
        
        @discardableResult public func property(_ customProperty: String, value: PropertyValue) -> Style {
            return property(.custom(customProperty), value: value)
        }
        
        internal func toString() -> String {
            let selectorsStr = selectors.map { $0.toString() }.joined(separator: ", ")
            var str = "\(selectorsStr) {\n"
            for prop in properties {
                str.append("\t" + prop.toString() + "\n")
            }
            str.append("}")
            
            return str
        }
    }
}


SwiftCSSBuilder/Classes/Selectors.swift
//
//
//  Selectors.swift
//
//  Created on 03/03/22
//  Copyright (c) 2022 Marco Siino. All rights reserved.
//


import Foundation

extension CSS {    
    public enum Selector {
        //Elements
        case all
        case html
        case body
        case p
        case a
        case div
        case span
        
        //Other
        case identifier(_ name: String)
        case `class`(_ name: String)
        case custom(_ selector: String)
        
        public func withClass(_ name: String) -> Selector {
            return .custom(self.toString() + "." + name)
        }
        
        public func withPseudoClass(_ name: String) -> Selector {
            return .custom(self.toString() + ":" + name)
        }
        
        public func withAttribute(_ name: String) -> Selector {
            return .custom(self.toString() + "[" + name + "]")
        }
        
        public func withAttribute(_ name: String, value: String) -> Selector {
            return .custom(self.toString() + "[\(name)=\(value)]")
        }
        
        public func insideOf(_ selector: Selector) -> Selector {
            return .custom(selector.toString() + " " + self.toString())
        }
        
        public func insideOf(_ name: String) -> Selector {
            return .custom(name + " " + self.toString())
        }
        
        public func childOf(_ selector: Selector) -> Selector {
            return .custom(selector.toString() + " > " + self.toString())
        }
        
        public func childOf(_ name: String) -> Selector {
            return .custom(name + " > " + self.toString())
        }
        
        func toString() -> String {
            switch self {
            case .all:
                return "*"
            case .html:
                return "html"
            case .body:
                return "body"
            case .p:
                return "p"
            case .a:
                return "a"
            case .div:
                return "div"
            case .span:
                return "span"
            case .identifier(let name):
                return "#\(name)"
            case .class(let name):
                return ".\(name)"
            case .custom(let selector):
                return selector
            }
        }
    }
}


SwiftCSSBuilder/Classes/Stylesheet.swift
//
//
//  Stylesheet.swift
//
//  Created on 03/03/22
//  Copyright (c) 2022 Marco Siino. All rights reserved.
//


import Foundation

extension CSS {
    public struct Stylesheet {
        private var styles = [Style]()
        
        public init(_ styles: [Style]) {
            self.styles = styles
        }
        
        public init() {
            self.styles = [Style]()
        }
        
        @discardableResult public func with(_ style: Style) -> Stylesheet {
            var styles = self.styles
            styles.append(style)
            return Stylesheet(styles)
        }
        
        public func toString() -> String {
            var str = ""
            for style in styles {
                str.append(style.toString() + "\n\n")
            }
            
            return str
        }
    }
}




Example/SwiftCSSBuilder/AppDelegate.swift
//
//  AppDelegate.swift
//  SwiftCSSBuilder
//
//  Created by Marco Siino on 03/08/2022.
//  Copyright (c) 2022 Marco Siino. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



SwiftCSSBuilder/Classes/UIColor+hex.swift
//
//
//  UIColor+hex.swift
//
//  Created on 03/03/22
//  Copyright (c) 2022 Marco Siino. All rights reserved.
//


import Foundation

extension UIColor {
    func toHex(withAlpha: Bool = true) -> String {
        guard let components = cgColor.components, components.count >= 3 else {
            if withAlpha {
                return "000000FF"
            }
            else {
                return "000000"
            }
        }
        
        let r = Float(components[0])
        let g = Float(components[1])
        let b = Float(components[2])
        
        if withAlpha == false {
            return String(format: "%02lX%02lX%02lX", lroundf(r * 255.0), lroundf(g * 255.0), lroundf(b * 255.0))
        }
        else {
            var a = Float(1.0)
            if components.count >= 4 {
                a = Float(components[3])
            }
            
            return String(format: "%02lX%02lX%02lX%02lX", lroundf(r * 255.0), lroundf(g * 255.0), lroundf(b * 255.0), lroundf(a * 255.0))
        }
    }
}


Example/SwiftCSSBuilder/ViewController.swift
//
//  ViewController.swift
//  SwiftCSSBuilder
//
//  Created by Marco Siino on 03/08/2022.
//  Copyright (c) 2022 Marco Siino. All rights reserved.
//

import UIKit
import SwiftCSSBuilder
import WebKit

class ViewController: UIViewController {

    @IBOutlet weak var webView: WKWebView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let style = CSS.Stylesheet()
            .with(CSS.Style(for: .all)
                    .property(.fontFamily, value: "Helvetica")
                    .property(.fontSize, value: .pt(20))
                    .property(.color, value: "#606060"))
            .with(CSS.Style(for: .class("cuteDiv"))
                    .property("background-color", value: "#eeeeee")
                    .property(.padding, value: .px(20)))
            .with(CSS.Style(for: .p.childOf(.div.withClass("cuteDiv")))
                    .property(.color, value: .color(.red)))
            .with(CSS.Style(for: .a)
                    .property(.color, value: .color(.blue)))
            .with(CSS.Style(for: .p)
                    .property(.marginTop, value: .em(1.5)))
        
        var htmlStr = "<style>\(style.toString())</style>"
        htmlStr.append("<p>This is a paragraph with <strong>bold text</strong>. This is <span>a span</span></p>")
        htmlStr.append("<p>This is another paragraph with <a href=\"https://www.google.it\">a link</a></p>")
        htmlStr.append("<div><p>This is paragraph inside a div</p></div>")
        htmlStr.append("<div class=\"cuteDiv\"><h1>Div header</h1><p>This is paragraph inside a div.cuteDiv</p></div>")
        webView.loadHTMLString(htmlStr, baseURL: nil)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

}



SwiftCSSBuilder/Classes/Property.swift
//
//
//  Property.swift
//
//  Created on 03/03/22
//  Copyright (c) 2022 Marco Siino. All rights reserved.
//

import Foundation

extension CSS {
    
    public enum PropertyType {
        case color
        case fontFamily
        case fontSize
        case lineSpacing
        case margin
        case marginTop
        case marginBottom
        case marginLeft
        case marginRight
        case padding
        case paddingTop
        case paddingBottom
        case paddingLeft
        case paddingRight
        case custom(_ name: String)
        
        func toString() -> String {
            switch self {
            case .color:
                return "color"
            case .fontFamily:
                return "font-family"
            case .fontSize:
                return "font-size"
            case .lineSpacing:
                return "line-spacing"
            case .margin:
                return "margin"
            case .marginTop:
                return "margin-top"
            case .marginBottom:
                return "margin-bottom"
            case .marginLeft:
                return "margin-left"
            case .marginRight:
                return "margin-right"
            case .padding:
                return "padding"
            case .paddingTop:
                return "padding-top"
            case .paddingBottom:
                return "padding-bottom"
            case .paddingLeft:
                return "padding-left"
            case .paddingRight:
                return "padding-right"
            case .custom(let name):
                return name
            }
        }
    }
    
    public enum PropertyValue {
        case custom(value: String)
        case px(_ value1: Int, _ value2: Int? = nil, _ value3: Int? = nil, _ value4: Int? = nil)
        case pt(_ value1: Double, _ value2: Double? = nil, _ value3: Double? = nil, _ value4: Double? = nil)
        case em(_ value1: Double, _ value2: Double? = nil, _ value3: Double? = nil, _ value4: Double? = nil)
        case perc(_ value1: Double, _ value2: Double? = nil, _ value3: Double? = nil, _ value4: Double? = nil)
        case color(_ value: UIColor)
        case RGB(_ red: Int, _ green: Int, _ blue: Int)
        case RGBA(_ red: Int, _ green: Int, _ blue: Int, _ alpha: Int)
        case fRGB(_ red: CGFloat, _ green: CGFloat, _ blue: CGFloat)
        case fRGBA(_ red: CGFloat, _ green: CGFloat, _ blue: CGFloat, _ alpha: CGFloat)
        case fontFamily(_ font: UIFont, fallbackFonts: [String]? = nil)
        case fontSize(_ font: UIFont)
        
        private func stringForValues<T>(withUnit unit: String, value1: T, value2: T? = nil, value3: T? = nil, value4: T? = nil) -> String {
            var str = "\(value1)\(unit)"
            if let value2 = value2 {
                str += " \(value2)\(unit)"
            }
            if let value3 = value3, let value4 = value4 {
                str += " \(value3)\(unit) \(value4)\(unit)"
            }
            return str
        }
        
        func toString() -> String {
            switch self {
            case .custom(let value):
                return value
            case .px(let value1, let value2, let value3, let value4):
                return stringForValues(withUnit: "px", value1: value1, value2: value2, value3: value3, value4: value4)
            case .pt(let value1, let value2, let value3, let value4):
                return stringForValues(withUnit: "pt", value1: value1, value2: value2, value3: value3, value4: value4)
            case .em(let value1, let value2, let value3, let value4):
                return stringForValues(withUnit: "em", value1: value1, value2: value2, value3: value3, value4: value4)
            case .perc(let value1, let value2, let value3, let value4):
                return stringForValues(withUnit: "%", value1: value1, value2: value2, value3: value3, value4: value4)
            case .color(let color):
                return "#\(color.toHex())"
            case .RGB(let r, let g, let b):
                let color = UIColor(red: CGFloat(r)/255.0, green: CGFloat(g)/255.0, blue: CGFloat(b)/255.0, alpha: 1.0)
                return "#\(color.toHex())"
            case .RGBA(let r, let g, let b, let a):
                let color = UIColor(red: CGFloat(r)/255.0, green: CGFloat(g)/255.0, blue: CGFloat(b)/255.0, alpha: CGFloat(a)/255.0)
                return "#\(color.toHex())"
            case .fRGB(let r, let g, let b):
                let color = UIColor(red: r, green: g, blue: b, alpha: 1.0)
                return "#\(color.toHex())"
            case .fRGBA(let r, let g, let b, let a):
                let color = UIColor(red: r, green: g, blue: b, alpha: a)
                return "#\(color.toHex())"
            case .fontFamily(let font, let fallbackFonts):
                var fonts = "\"\(font.fontName)\""
                if let fallbackFonts = fallbackFonts {
                    for font in fallbackFonts {
                        fonts.append(", \(font)")
                    }
                }
                return fonts
            case .fontSize(let font):
                return "\(font.pointSize)pt"
            }
        }
    }
    
    public struct Property {
        private var property: PropertyType
        private var value: PropertyValue
        
        public init(property: PropertyType, value: PropertyValue) {
            self.property = property
            self.value = value
        }
        
        internal func toString() -> String {
            return "\(property.toString()): \(value.toString());"
        }
    }
}
